### ISSUE TITLE
Aiming for July 1st for Godot 3.1 Alpha

### ISSUE THREAD

**Issue by reduz** [godotengine/godot#19101]
We were discussing a date we can aim for in order to release the first alpha for Godot 3.1. As most of the proposed features are getting completed nicely, July 1st sounds like a reasonable date for releasing the first alpha for this cycle.

Major features mostly complete (let me know if i'm forgetting anything!)

-2D Meshes, 2D Skeletons, CSG and new Inspector (me)
-Physical Bones (@AndreaCatania)
-Input Strength system (@groud)
-GDNative C++ API rework (@karroffel)

Major features currently in development:

-OpenGL ES2 backend (@karroffel)
-Typed GDScript (@vnen)
-New animation editor, animation tree editor and visual shader editor (me)
-IK and softbodies (@AndreaCatania)
-Editor theme improvements (@djrm)

Again, let me know if I forgot anyone!

The idea is to have these features mostly impemented (have to be more or less working) for July 1st alpha. Then we can aim for beta, candidates and release sometime in August, at much September.

What do you guys think?

**Comment by BastiaanOlij** [comment]
Sounds good to me

**Comment by frank-lesser** [comment]
just to mention @Calinou 's Sponza demo crashed in the latest master ( worked ok in commit 17e8456 v. 12.5 ) - have also memory crashes in my latest synch

**Comment by volzhs** [comment]
#18865 new android export system?

**Comment by reduz** [comment]
@frank-lesser this is for discussing features related to alpha, please open a separate issue for bugs and enhancements

**Comment by uldall** [comment]
I believe there was talk about 3.1 being the release for getting C# production-ready. #18364

**Comment by reduz** [comment]
@uldall That depends on @neikeq 's schedule. As I understand, he is currently busy until July finishing an internship as part of his education. Afterwards we are aiming to be able to hire him to focus on completing C# support. That said I think exporting has progressed a good amount since 3.0 came out.

**Comment by mhilbrunner** [comment]
July 1st sounds great!

**Comment by eon-s** [comment]
An updated roadmap to see what is still missing and what will be kicked for 3.2?

**Comment by Megalomaniak** [comment]
So I take it the ShaderGraph is slated for a come-back some time after 3.1? I'm asking here since many in the forums have been asking about it.

**Comment by vnen** [comment]
@Megalomaniak as I understand, the visual shader editor will come in 3.1 (will be present in this alpha).

**Comment by Megalomaniak** [comment]
That is great to hear. :)

**Comment by swarnimarun** [comment]
I don't want to but New animation editor, animation tree editor and visual shader editor seems like a lot of work for one person. Best of luck reduz.

**Comment by vnen** [comment]
> I don't want to but New animation editor, animation tree editor and visual shader editor seems like a lot of work for one person. Hope everything will be fine.

It's not a regular person, it's @reduz :wink:

**Comment by samvila** [comment]
@reduz Any estimation of when are we going to have new 3D capabilities? I'm missing:

- Occlusion culling (per face/per vertex would be even better than per object but even this last one would be amazing to have it)
- Motion Blur for camera and for objects
- Decals
- Soft shadows

Some of these features were announced for 3.1 if I'm not mistaken.

**Comment by aaronfranke** [comment]
I don't know about advertising C# as production-ready yet. There have been many changes since 3.0, and I think we won't have too many more breaking ones after 3.1, but C# support still has some issues.

Can we decide on a milestone for #5813 so that it's not forgotten again? Either 3.1 or 4.0.

If the Vulkan renderer will be included in Godot 3.2, then it would probably make sense for #288 to be included in 3.2 as well, because Vulkan will be the first renderer we've used to support doubles.

If possible I'd like to do some math refactoring before 3.1 is out, right now I'm waiting on a PR for moving Vector2 to its own file #18096

**Comment by mhilbrunner** [comment]
Thing that I'd expect before I'd consider C# *production ready*, in order of importance:
- Crashes ([there are ~25 known ones currently](https://github.com/godotengine/godot/issues?utf8=%E2%9C%93&q=is%3Aissue+is%3Aopen+mono+label%3Atopic%3Amono+crash))
- Exporting (https://github.com/godotengine/godot/issues/15615)
- API should be mostly final. There are still some discussion that don't appear resolved? (https://github.com/godotengine/godot/issues/14466)
- Editing, renaming, (re)moving files should be supported; there should be no need to hand-edit generated project files (https://github.com/godotengine/godot/issues/12917, https://github.com/godotengine/godot/issues/12415, https://github.com/godotengine/godot/issues/11930)
- Debugging should work (https://github.com/godotengine/godot/issues/13837, https://github.com/godotengine/godot/issues/14822)
- We should have at least one minor version (so 3.1) with working exporting for C# supported before we can declare the exporter *production ready*

**Comment by KellyThomas** [comment]
> API should be mostly final.

Further to this point there are public elements that don't conform to the newly written style guide yet.

**Comment by MCrafterzz** [comment]
what is a 2d mesh?

**Comment by akien-mga** [comment]
> Things that I'd expect before I'd consider C# production ready, in order of importance:

That should likely go into its own dedicated tracker for further discussion.

**Comment by swarnimarun** [comment]
@MCrafterzz 2D Meshes are what allow deformation of sprites. That's what they are in lay-man's terms.
But they will be considered a 2d mesh hull for the sprite using which mesh like alterations can be performed onto the sprites. 2D skeletal uses this, I believe.

**Comment by neikeq** [comment]
@mhilbrunner I made a list in #18364 and asked for comments if someone thinks there is something missing, so mention those you think are missing there.

> - API should be mostly final. There is still some discussion that doesn't appear resolved? (https://github.com/godotengine/godot/issues/14466)

I don't think any of the changes from #14466 will happen, at least not in 3.x. We didn't reach a consensus so the best is to stay with the current names. There are other pending API changes that need to be addressed, as I mention in #18364.

> - Debugging should work (https://github.com/godotengine/godot/issues/13837, https://github.com/godotengine/godot/issues/14822)

No, this will happen after 3.1 together with IDE extensions. Debugging is very important, but not required for it to be production ready. It's also not a simple task that can be finished quickly. #14822 was already fixed.

> - We should have at least one minor version (so 3.1) with working exporting for C# supported before we can declare the exporter *production ready*

Not sure I understood this, or how this differs from your second point.

**Comment by neikeq** [comment]
@aaronfranke Please comment in #18364 what you think is needed.

Again, let's move the C# towards 3.1 topic there, or a tracker issue if desired.

**Comment by mhilbrunner** [comment]
@neikeq Answering in #18364 to move the discussion out of this issue.

**Comment by ca3games** [comment]
When will the google summer of code features will be released? I really want the MIDI input.

**Comment by vnen** [comment]
@ca3games GSOC ends in August. It's much too late to make into 3.1.

**Comment by eon-s** [comment]
@vnen alpha means feature freeze then?

**Comment by MCrafterzz** [comment]
Will https://github.com/godotengine/godot/pull/15110 make it into 3.1?

**Comment by swarnimarun** [comment]
@MCrafterzz It's on the 3.1 milestone and it works perfectly. Only thing left is the commits to be squashed and another review by the devs. So it will more than likely make it in.

**Comment by vnen** [comment]
> alpha means feature freeze then?

Yes. Alpha means feature-complete. Only bug-fixes and small enhancements will go in after that. New features will wait until the next release cycle.

**Comment by mhilbrunner** [comment]
Some things that were not mentioned in the issue that were still to be discussed for 3.1 (should we move those to 3.2 instead?):

- Multiple camera support (https://github.com/godotengine/godot/pull/8215, https://github.com/godotengine/godot/pull/10950)
- Debugger rework (https://github.com/godotengine/godot/pull/8644 and lots of others that were postponed until after this rework)
- Renderer features (will likely need to wait for Vulkan, so 3.2?) -> Occlusion culling, stencil support, decals & so on

**Comment by swarnimarun** [comment]
@mhilbrunner I don't know about the rest but in a few issues I have seen @reduz answer that the renderer features will only be added with 3.2(Vulkan Support).

Multi-camera support will really be awesome though. Probably also accessibility for layer-masking in camera. And ability to get_texture from camera. But it can wait. Man, I am just too greedy.

**Comment by LinuxUserGD** [comment]
Renderer features / Shader caching:
Will Godot still keep OpenGL 3.3 / OpenGL ES 3.0 compatibility (Mesa driver fallback) or switch to a higher OpenGL version in the future (using Mesa core profile version)?

**Comment by akien-mga** [comment]
>  Will Godot still keep OpenGL 3.3 / OpenGL ES 3.0 compatibility (Mesa driver fallback) or switch to a higher OpenGL version in the future (using Mesa core profile version)?

Keep OpenGL 2.1 and OpenGL 3.3 compatibility (gles2 and gles3 renderers) for a while (as long as not 99% of devices support Vulkan), defaulting to Vulkan where available. We won't go above OpenGL 3.3 / OpenGL ES 3.0 a priori as far as GL is concerned.

**Comment by swarnimarun** [comment]
@MCrafterzz I recommend a quick google search but if you are lazy,
See this PR. https://github.com/godotengine/godot/pull/19264

**Comment by Ranoller** [comment]
2D meshes are very promising, i´m testing extensively (some bugs jet, when have some time i will report)... but this feature makes the difference in 2D world... so please think in the **internal** triangulation visibility and edition of the polygon (or the mesh) because this is necesary to achieve a lot of the effects that a tool like this is intended to achieve...

**Comment by akien-mga** [comment]
> We were discussing a date we can aim for in order to release the first alpha for Godot 3.1. As most of the proposed features are getting completed nicely, July 1st sounds like a reasonable date for releasing the first alpha for this cycle.

For those who might wonder where the Godot 3.1 alpha builds are tomorrow: don't wait for them :) It's not ready yet, and the July 1st date was just a tentative direction for the project, not a deadline. Things are shaping up well, but I think there's at least a couple weeks more before we can release an alpha.

**Comment by definite-d** [comment]
Finally. Tomorrow is the day!

**Comment by Ranoller** [comment]
Did you read last message?

**Comment by definite-d** [comment]
@Ranoller I meant that from tomorrow we could begin expecting a 3.1 release. :)

**Comment by eon-s** [comment]
I think it would be nice to start getting the big features (typed gd, soft bodies, gles2) to have more time for testing before alpha.

**Comment by MCrafterzz** [comment]
Has a alpha been released?

**Comment by groud** [comment]
> Has a alpha been released?

Just... stop asking. 
Don't worry, when alpha is released you'll know it. We will communication on all social media about it.

**Comment by eon-s** [comment]
@MCrafterzz main big features are already in and if you look the commits, looks like devs are ironing out the details on some parts and fixing old bugs, so alpha is close, just wait a bit more.

**Comment by dutchmo** [comment]
No need to wait -- I've been playing around with the new features for a couple of weeks. Just download a daily build, if you are on linux, its chmod +x, and run.

Or am i missing something? Alpha just means "we're going to call this build alpha and let everyone know".  A lot of stuff can be done already.

But if you are doing production games, you'll not be wanting alpha anyway, in most cases.

**Comment by akien-mga** [comment]
Closing as superseded by #18364.

> Or am i missing something? Alpha just means "we're going to call this build alpha and let everyone know". A lot of stuff can be done already.

Yes, the alpha 1 will just be a snapshot of the master branch, like nightly builds. It will just be more "official", easier to find for the masses, and will likely get broader testing.

**Comment by definite-d** [comment]
Alpha already! This is exciting.
